File: Data Structures & Algorithms/count-subsequences/submission-1.cs
public class Solution {
    private int?[,] dp;

    public int NumDistinct(string s, string t) {
        int m = s.Length, n = t.Length;
        if (n > m) return 0;
        dp = new int?[m + 1, n + 1];
        return Dfs(s, t, 0, 0);
    }

    private int Dfs(string s, string t, int i, int j) {
        if (j == t.Length) return 1;
        if (i == s.Length) return 0;
        if (dp[i, j].HasValue) return dp[i, j].Value;

        int res = Dfs(s, t, i + 1, j);
        if (s[i] == t[j]) {
            res += Dfs(s, t, i + 1, j + 1);
        }
        dp[i, j] = res;
        return res;
    }
}

File: Data Structures & Algorithms/swim-in-rising-water/submission-1.java
public class Solution {
    public int swimInWater(int[][] grid) {
        int n = grid.length;
        boolean[][] visit = new boolean[n][n];

        return dfs(grid, visit, 0, 0, 0);
    }

    private int dfs(int[][] grid, boolean[][] visit, int r, int c, int t) {
        int n = grid.length;
        if (r < 0 || c < 0 || r >= n || c >= n || visit[r][c]) {
            return 1000000;
        }
        if (r == n - 1 && c == n - 1) {
            return Math.max(t, grid[r][c]);
        }
        visit[r][c] = true;
        t = Math.max(t, grid[r][c]);
        int res = Math.min(Math.min(dfs(grid, visit, r + 1, c, t),
                                     dfs(grid, visit, r - 1, c, t)),
                           Math.min(dfs(grid, visit, r, c + 1, t),
                                    dfs(grid, visit, r, c - 1, t)));
        visit[r][c] = false;
        return res;
    }
}